### PR TITLE
chore: Add a metadata provider to handoff IMDS calls

### PR DESCRIPTION
**Pull request by jonathan-innis**
Fixes # 

**Description**

- Adds a metadata provider to handoff metadata resolution in cloudprovider
- Adds a generic caching mechanism for strings
- This PR is a pre-req to the larger NTH changes

**How was this change tested?**

* `make battletest`

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: # 
- [x] No

**Release Note**

```release-note
None
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/cloudprovider/aws/cloudprovider.go [hunk 1]
```diff
@@ -95,11 +95,8 @@ func NewCloudProvider(ctx context.Context, options cloudprovider.Options) *Cloud
 			client.DefaultRetryer{NumMaxRetries: client.DefaultRetryerMaxNumRetries},
 		),
 	)))
-	if *sess.Config.Region == "" {
-		logging.FromContext(ctx).Debug("AWS region not configured, asking EC2 Instance Metadata Service")
-		*sess.Config.Region = getRegionFromIMDS(sess)
-	}
-	logging.FromContext(ctx).Debugf("Using AWS region %s", *sess.Config.Region)
+	metadataProvider := NewMetadataProvider(sess, NewEC2MetadataClient(sess), sts.New(sess))
+	metadataProvider.MustEnsureRegion(ctx, sess) // resolves the region in the session config
```
Comment (ellistarn) [inline pkg/cloudprovider/aws/cloudprovider.go:99]: IMO we should just resolve this in the constructor (or panic)

File: pkg/utils/ptr/ptr.go [hunk 2]
```diff
@@ -30,3 +30,7 @@ func Node(node v1.Node) *v1.Node {
 func Quantity(quantity resource.Quantity) *resource.Quantity {
 	return &quantity
 }
+
+func Val[T any](v T) *T {
```
Comment (ellistarn) [inline pkg/utils/ptr/ptr.go:34]:
```suggestion
func To[T any](v T) *T {
```
Idiomatically:
```
ptr.To(mything)
```

Comment (ellistarn) [reply pkg/utils/ptr/ptr.go:34]: Also, this use of generics is awesome and we should replace all other ptr methods w/ it. Maybe we need a corresponding `From`.

**Comment by jonathan-innis** [comment]
Closing in favor of adding some components to the NTH PR

**Pull request closed without merging**